MDEV-11670: systemd multi-instance

grooverdan opened the pull request:
The multi instance system configuration was a bit of a mess when added, sorry. It was effectively --defaults-file=@sysconf2dir@/my%I.cnf where each instance name had a new file.

As @sysconfdir@/my.cnf included these all files it was perhaps a bad choice as an existing single instance mariadb server would include all the multi-instance configuration files.

As 10.3 isn't EOL ~~GA~~ ~~or Beta~~ yet, it seems a good time to revise this back to a saner default. Those using the existing configuration have an easy migration plan and a mechanism to prevent the incorrect starting is introduced.

Requires #509 and #508 for correct mysqld_install_db behaviour.

Edit: merged the #387 for multi-instance here.

I submit this under the MCA.

Distro people, @ottok, @oerdnj any improvements/problems?

I was almost contemplating Command Line Mechanism as the default however this could imply some maintenance burden on directories/selinux rules etc.

grooverdan committed:
systemd: multi-instance changes to -defaults-group-suffix=.%I

When the multi-instance systemd service file was chosen it effectively
relied on /etc/my.cnf.d/my{instancename}.cnf file to define its
configuration file. This is problematic if running along side a
single instance mariadb service which has /etc/my.cnf that reads all
configuration file /etc/my.cnf.d/*.cnf.

To prevent the service from auto starting up if a user has this
previous configuration ConditionPathExists=!@sysconf2dir@/my%I.cnf
to ensure that a user with the previous configuration isn't
started in a non-intended mode. Documentation in the service file
(should be release notes too), described a recommended migration.

A new approach was to use --defaults-group-suffix=.%I as an
arguement to mysqld and let the user define a [mysqld.{instancename}]
group within the configuration file. This way existing global
mysqld configuration options are read with the instance name
having special overrides of datadir, port, socket etc.

A systemd environment variable MYSQLD_MULTI_INSTANCE is used in the
defination as it give the user flexability to use multiple
segregation mechanisms between services. This is used multiple
times within the service which all needed to be kept consistent.

Another notable change is mysql_install_db being part of the
ExecStartPre. This provides and auto-initialization for users
that run multiple instances.

grooverdan commented:
Rebased again. As this is a non-stable change it would be good to merge this pre-GA.

codership-jenkins commented:
Can one of the admins verify this patch?

grooverdan commented:
The default mysql_install_db  is the only thing I'm less sure I've included it in the multiinstance as its harmess on an existing populated datadir as I understand it and which multiinstances can occur aren't predefined and therefore cannot be managed by packaging.

ProCycleDev commented:
The bugs in the current multi-instance service file that I have encountered have all been fixed here.

The Aliases being defined in this service file template breaks the backwards compatibility symlinks which the preinstall script depends on (makes updates fail in apt) because it tries to stop/start mysql.service which no longer functions because the symlinks now point to mariadb@.service instead of the expected mariadb.service

Moving the location of the instance configuration files fixes another bug where the default my.cnf file loads all of the instance files in conf.d/ which overwrites the default server block and causes the service to fail to start because the socket/files are already in use. This also breaks upgrades due to the non functional default mysql.service blocking the upgrade and leaving the package in a broken state.

grooverdan commented on support-files/mariadb@.service.in:
```diff
@@ -1,35 +1,116 @@
+#
+# Configuration File Based Mechanism:
+#
+# MYSQLD_MULTI_INSTANCE=--defaults-file=@sysconfdir@/my%I.cnf
```
https://jira.mariadb.org/browse/MDEV-11869?focusedCommentId=90993&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-90993

Note in above, @ottok wanted ```@sysconfdir@mariadb.%I.cnf``` which is a trivial docs change.

grooverdan commented:
:disappointed: :crying_cat_face:

ottok commented:
@fauust Do you have any systemd insights here?

grooverdan commented:
@ottok, there's enough insights from users like @ProCycleDev, me and @eworm-de who have systemd knowledge that actually have encountered this problem. It seems @fauust has done some very basic debian packaging however deferring however yet another person on the bug report isn't going to help this.

Stop absconding from packaging responsibility (over a #&$@ year!) and can somebody actually spend 2 minutes looking though the changes, ask some questions and then try testing it and then merge.

If you need some tests around packaging/systemd point to a repository and I'll contribute some.

Except for a few favoured business partners the merge rate of PRs is pathetic.

fauust commented:
@grooverdan I did not had time to look deeper in this issue, sorry for that. And still I am not able to validate this MR. I was working lately on MDEV-15526 (currently building and testing a fix). The only thing I am sure of is that using Alias is not recommended.
See: https://lists.freedesktop.org/archives/systemd-devel/2018-November/041658.html and https://wiki.debian.org/Teams/pkg-systemd/Packaging#systemd_unit_files_naming_and_installation that was updated after my question on the ML.

Hope that helps!

fauust commented:
FYI this is the fix I am currently testing https://github.com/fauust/mariadb/commit/48358cca9f6ae0528311e7755b79af4c2ca370a5

grooverdan commented:
@fauust that's nice but large irrelevant to this PR (except I just happened to notice aliases were crap a while ago) so it doesn't help. Why its considered sane to install sysv and systemd files at the same time is very troubling (when there are rpm/deb flags to check at install time right?) (like the highly unpopular, unfixed, ignored MDEV-10797). What functionality symlinks hopes to bring to this mess is also suspect as a potential solution (because its on a wiki page doesn't make it sane).

back to this PR:

My gripe isn't with @fauust, its with the lack of interest in looking at submitted work that users (INCLUDING THE CONTRIBUTOR) have noticed and a continued bouncing/ignoring of the issue (like so many PRs).

To test this, build and install it, try setting MYSQLD_MULTI_INSTANCE option as per the file and start multiple instances like `systemctl start mariadb@bob`, `systemctl start mariadb@julie` and see if you have two separate instances. After that, clear those instances out and try another MYSQLD_MULTI_INSTANCE mechanism.

I'll happily remove the galera parts (it will make testing easier right?) to this if anyone wants, which, with small modifications (remove `PermissionsStartOnly=true`, remove galera bits, add `--user=mysql` to default `MYSQLD_MULTI_INSTANCE` + doc cleanups) will be able to run a database instance per user. A separate `mariadb_galera@` could be created if there is demand.

So please, ask questions.

ottok commented:
We really appreciate your numerous contributions. Unfortunately due to lack of manpower to review, and manpower to follow-up, test and fix in case of regressions, and lack of good CI support to due some of this work automatically there is often the case that while people do comment, they cannot ultimately commit all the way.

I know from Zulip that Faustin is at least testing this next week once the Debian activities calm down.

grooverdan committed:
systemd: multi-instance not for Galera, User/Group flexible

By removing Galera functionality, we remove PermissionsStartOnly=true
and hence make this service more flexible for running multiple
instances each on a different user.

grooverdan commented:
Rebased on 10.4 and removed Galera component of multi-instance for testing ease and added ability to change the MySQL user easier. Also make it much more readable. Separated fixes into separate PRs against 10.1 branch.

fauust commented:
@grooverdan sorry for the delay, I am back to systemd testing and will get back to you soon regarding this and also MDEV-15526.

fauust commented:
@grooverdan, I am able build and install it from your branch, dependence on galera seems not problematic for my tests: galera-3 can be installed from debian stretch repo and galera-4 from official mariadb repo.

Are we OK that this is the 10.4 branch (even if your branch name contains 10.3 that seems to be related with the age of your PR?).

I will be testing this on debian stretch, if you have anything new you want me to test, let me know please.

grooverdan commented:
10.4 is fine and is the currently targeted branch from the last cleanup/rebase I did.

Galera SST mechanisms and their interaction with systemd is where these tests get complicated.
#1143 has a much nicer basis for interaction and, if there is a demand, I can look at using similar concepts in a galera multinstance however I see the multiinstance as more for shared hosting and multiversion testing which is slightly different from what galera is normally used for.

For tests I suggest trying out a few of the multiinstance scenarios listed in the top of the file.

Also test upgrade paths if a user has a multiinstance and not a main mariadb.service instance to see if the scripts behave in a sane way (@ProCycleDev 's comments above - https://github.com/MariaDB/server/blob/10.4/debian/mariadb-server-10.4.postinst#L279 looks like it will start if service disabled). Don't try to be too smart in the packaging scripts regarding multiservice, especially do nothing with user account maintenance. A small notice to say multinstance services may need restart is probably sufficient. Anything more would need to ensure the service map to the mysqld that was upgraded.

fauust commented on support-files/mariadb@.service.in:
```diff
@@ -1,29 +1,120 @@
+#
+# Configuration File Based Mechanism:
+#
+# MYSQLD_MULTI_INSTANCE=@sysconfdir@/my%I.cnf
```
Shouldn't it be?
`MYSQLD_MULTI_INSTANCE="--defaults-file=@sysconfdir@/my%I.cnf"`

fauust replied:
This seems to be already noticed above (https://github.com/MariaDB/server/pull/510#issuecomment-390801929)

fauust commented on support-files/mariadb@.service.in:
```diff
@@ -140,6 +197,19 @@ UMask=007
 PrivateTmp=false
 
+# Controlling how multiple instances are separated. See top of this file.
+# Note 1: This service isn't User=mysql be default so we need to be explicit.
```
Not sure I understand exactly this Note.

fauust commented:
@grooverdan, here are some questions, observations:
- Aliases will be removed from unit (see MDEV-15526 and https://github.com/MariaDB/server/pull/1172). In the case of multiple instance users are perfectly aware of what systemctl command they should use so it seems even less necessary to do the mysql|d aliases.
- I think we should leave the following test condition that shows immediately that configuration file is missing to user through `journalctl -xe`:
`ConditionPathExists=/etc/mysql/my%I.cnf`

fauust reviewed (changes requested):
@grooverdan can you please also check the following

fauust commented on support-files/mariadb@.service.in:
```diff
@@ -1,29 +1,120 @@
+# created these file(s).
+#
+## See Environment=MYSQLD_MULTI_INSTANCE below for current recommended options.
+ConditionPathExists=!@sysconf2dir@/my%I.cnf
```
```suggestion
ConditionPathExists=!@sysconf2dir@/my%I.cnf
ConditionPathExists=/etc/mysql/my%I.cnf
```

fauust reviewed (changes requested):
@grooverdan this looks ok to me. I am able to:
- run multiple instance;
- make these instance persistent to a reboot;
- enable or disable them;
- upgrade to the latest MariaDB version.

I'll have to update the documentation with the new path for the file conf (/etc/mysql/my%I.cnf) https://mariadb.com/kb/en/library/systemd/#interacting-with-multiple-mariadb-server-processes.

grooverdan committed:
MDEV-11670: ensure sysconfdir/sysconf2dir are not blank for mariadb@.service

grooverdan committed:
MDEV-11670: mariadb@.service remove alias, clean up documentation/order

grooverdan commented:
Thanks for the review and test @fauust.

I've added commits to this branch to address comments and one fix.

First commit enables a non-RPM/DEB build to contain the sysconf/sysconf2dir to be populated. Within the support files the mariadb@ instance is the only user.

Quite right - remove aliases here.

The config file multi instance mechanism isn't the default, the group suffix mechanism is. In addition to cleaning up the typos, I've grouped examples of multinstance together adding `ConditionPathExists` to the Configuration File Mechanism.

fauust commented:
Ok, perfect!
I will build and test this today and then I hope we will finally merge this ;-)

fauust commented on support-files/mariadb@.service.in:
```diff
@@ -1,38 +1,155 @@
+# controls each instance to ensure it is run independently. It is passed to
+# mysqld and mysql_install
+#
+# By default, a group suffix exists andw ithin the default configuration
```
```suggestion
# By default, a group suffix exists and within the default configuration
```

fauust commented:
@grooverdan, building and install is OK but I am wondering if /etc/my%I.cnf is the right PATH (given by `@sysconfdir@/my%I.cnf`)?

Shouldn't that be `/etc/mysql/my%I.cnf` -> (`@sysconfdir@/mysql/my%I.cnf`)?

Ah and there is a small typo error, see above.

fauust committed:
fix typo support-files/mariadb@.service.in

Co-Authored-By: grooverdan <daniel@linux.ibm.com>

grooverdan commented:
It is currently correct. RPMs don't use /etc/mysql and only debian does. If Debian wants to use /etc/mysql then it should set that in the layout to INSTALL_SYSCONFDIR_DEB.

fauust commented:
OK I understand now, I will check with @ottok and @cvicentiu for this to be merged.
Thanks again @grooverdan !

ottok commented:
@fauust I no comments on this one. I don't understand the systemd domain enough. Whoever understands this should be the person to review/merge, and I guess that means you

grooverdan commented:
> I hope we will finally merge this ;-)

Me too. What's the delay?

hgxl64 commented:
pushed into 10.4

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MariaDB/server?pullRequest=510) <br/>All committers have signed the CLA.

hgxl64 commented:
closing again. PR was accepted.
